libflux/src/core/parser/mod.rs

```diff
@@ -591,6 +593,82 @@ impl Parser {
         }));
     }
 
+    #[cfg(test)]
+    // "(" [Parameters] ")" "=>" MonoType
+    fn parse_function(&mut self) -> MonoType {
+        let _lparen = self.expect(TOK_LPAREN);
+
+        let mut params = Vec::<ParameterType>::new();
+        if self.peek().tok == TOK_PIPE_RECEIVE
+            || self.peek().tok == TOK_QUESTION_MARK
+            || self.peek().tok == TOK_IDENT
+        {
+            params = self.parse_parameters();
+        }
+        let _rparen = self.expect(TOK_RPAREN);
+        self.expect(TOK_ARROW);
+        let mt = self.parse_monotype();
+        if params.len() == 0 {
+            return MonoType::Function(Box::new(FunctionType {
+                base: self.base_node_from_other_end(&_lparen, &base_from_monotype(&mt)),
+                parameters: None,
+                monotype: mt,
+            }));
+        }
+        return MonoType::Function(Box::new(FunctionType {
+            base: self.base_node_from_other_end(&_lparen, &base_from_monotype(&mt)),
+            parameters: Some(params),
+            monotype: mt,
+        }));
+    }
+
+    #[cfg(test)]
+    // Parameters = Parameter { "," Parameter } .
+    fn parse_parameters(&mut self) -> Vec<ParameterType> {
+        let mut params = Vec::<ParameterType>::new();
+        let parameter = self.parse_parameter_type();
+        params.push(parameter);
+        while self.peek().tok == TOK_COMMA {
+            self.consume();
+            let parameter = self.parse_parameter_type();
+            params.push(parameter);
+        }
+        return params;
+    }
+
+    #[cfg(test)]
+    // [ "<-" | "?" ] identifier ":" MonoType
+    fn parse_parameter_type(&mut self) -> ParameterType {
+        let id;
+        let mut start = None;
+        if self.peek().tok == TOK_IDENT {
+            id = self.parse_identifier();
+        } else if self.peek().tok == TOK_PIPE_RECEIVE {
+            start = Some(self.expect(TOK_PIPE_RECEIVE));
+            id = self.parse_identifier();
+        } else if self.peek().tok == TOK_QUESTION_MARK {
+            start = Some(self.expect(TOK_QUESTION_MARK));
+            id = self.parse_identifier();
+        } else {
+            id = self.parse_identifier();
+        }
```

Comment:
Looks like a good place for a match expression
```Rust
match self.peek().tok {
    TOK_IDENT => {
        id = self.parse_identifier();
        ...
    },
    TOK_PIPE_RECEIVE => {
        self.consume();
        ...
    },
    TOK_QUESTION_MARK => {
        self.consume();
        id = self.parse_identifier();
        ...
    },
    _ => ...,
}
```

Reply:
✅